### ISSUE TITLE
v1 - Bring us up to spec

### ISSUE THREAD

**Issue by tim-phillips** [possibilities/micro-cors#51]
In order to bring this library up to the CORS spec, we need to accomplish the following tasks.

- [x] #48 

Anything else?

This will go towards a v1 release.

Please review the spec and discuss here. 

- https://www.html5rocks.com/en/tutorials/cors/
- https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS

cc @bukinoshita @infernalmaster @lemol

**Comment by tim-phillips** [comment]
What do you all think about the `Vary` header? Is that something we should handle or leave up to the user?

> If the server specifies a single origin rather than the "*" wildcard, then the server should also include Origin in the Vary response header — to indicate to clients that server responses will differ based on the value of the Origin request header.

https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#Access-Control-Allow-Origin

**Comment by metamas** [comment]
@tim-phillips Seems like ya'll lost steam on this. I'd like to help. I've recently started building a Next project and will need to handle some CORS stuff in my API Routes (which use `micro`). Spent a few days studying the MDN docs on CORS. Feel like I have a solid understanding of it all.

If I start working on some PRs, will you be available to review and help clarify anything I'm confused about?

**Comment by tim-phillips** [comment]
@metamas Yes! 

We lost steam because the steps are not clear right now. We need to review the spec and create actionable issues to bring this lib to spec. 

It can be useful before you create a PR to create an Issue to discuss what you're planning, or we can chat about it in this thread if you think that's a better place for it.

**Comment by metamas** [comment]
@tim-phillips Ah, I see. Is the lack of clarity with the CORS spec, or just with how to implement it for `micro`?

My primary interest is to have CORS middleware to use with my `next` project. I'm doing all the server-side stuff with the newer API Routes feature that uses micro for route handlers. I figured helping to get this project up to snuff is the  efficient option.

But if the requirement to do that aren't clear, then I'm not sure.

**Comment by tim-phillips** [comment]
@metamas Yes, precisely. This lib does work fine, it just doesn’t follow the spec exactly.

**Comment by metamas** [comment]
@tim-phillips I'll fork and approach this by simply reading through the spec. As I do so, I'll check that the project has tests for specified behavior. If it does, I'll check their validity. If it doesn't, I'll add tests and work on getting them to pass (if they don't already). I'll also return here to add to this list.

Ok. So that's my intention. Now, I just need to manage enough free-time to get it done. 😅

Any suggestions on my approach?

**Comment by tim-phillips** [comment]
@ramiel Added to OP. While allow credentials is true, do you know if the client/browser is responsible for rejecting when the `Origin` is `*`?

@metamas I like the test driven approach.

**Comment by pudgereyem** [comment]
I opened up a PR with some changes to the `v1` branch that you can see here; https://github.com/possibilities/micro-cors/pull/76

Appreciate anyone taking the time to look it through. Keen on getting `v1` in a shape to get merged to `master`.
